### PR TITLE
feat: add services restart to `Set-EsxiCertificateMode`

### LINKED ISSUE #33: Enhance `Set-EsxiCertificateMode` to restart vCenter Server services
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Description

Enhance `Set-EsxiCertificateMode` to restart vCenter Servers services after the mode is changed.

Currently, restarting vCenter Server services is a manual process. 

### Use Case(s)

NA

### Potential Configuration

NA

### References

_No response_

### PR DESCRIPTION
**Pull request by garlicNova**
### Summary

Added a restart for specific vCenter Server services to `Set-EsxiCertificateMode` function.

### Type

- [ ] Bugfix
- [x] Enhancement or Feature
- [ ] Code Style or Formatting
- [ ] Documentation
- [ ] Refactoring
- [ ] Chore
- [ ] Other
        Please describe:

### Breaking Changes?

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### Test and Documentation

- [x] Tests have been completed.
- [ ] Documentation has been added or updated.

### Issue References

Closes #33

### CODE REVIEW

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Please amend with an update to the `CHANGELOG.md` and a bump to the build in the `.psd1`.

Example:

```markdown

## v1.5.2

> Release Date: Unreleased

Enhancement:

- Added a restart for specific vCenter Server services to `Set-EsxiCertificateMode` function. [GH-110](https://github.com/vmware/powershell-module-for-vmware-cloud-foundation-certificate-management/pull/110)

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
What about something like this:

```
function Restart-And-CheckService($serviceName) {
    Invoke-RestartService -Service $serviceName
    Start-Sleep -Seconds 30
    $serviceStatus = Invoke-GetService -Service $serviceName
    if ($serviceStatus.state -eq "STARTED") {
        Write-Output "$serviceName restarted successfully."
        return $null
    } else {
        Write-Error "$serviceName failed to restart."
        return $serviceName
    }
}
```

Then:

```powershell
$services = @("certificateauthority", "certificatemanagement")
$failedServices = @()

foreach ($service in $services) {
    $failedService = Restart-And-CheckService $service
    if ($null -ne $failedService) {
        $failedServices += $failedService
    }
}

if ($failedServices.Count -gt 0) {
    $failedServicesList = [string]::Join(", ", $failedServices)
    Write-Error "The following services failed to start: $failedServicesList. Operation Failed." -ErrorAction Stop
}
```

But perhaps add / modify the sleep with a counter?

And then add a check to verify that the mode is set as desired.

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 1]
```diff
@@ -612,8 +612,16 @@ Function Set-EsxiCertificateMode {
         $certModeSetting = Get-AdvancedSetting "vpxd.certmgmt.mode" -Entity $vCenterServer.connection
         if ($certModeSetting.value -ne $mode) {
             Set-AdvancedSetting $certModeSetting -Value $mode -confirm:$false
-            Write-Output "Certificate Management Mode is set to $mode on the vCenter Server instance $($vCenterServer.details.fqdn)."
-            Write-Output "Please restart the vCenter Server services for the change to take effect. See the vCenter Server Configuration documentation for information about restarting services."
+            # Restart "VMware Certificate Authority Service" and "VMware Certificate Management Service"
+            Invoke-RestartService -Service "certificateauthority"
```
Comment (bhumitra) [inline VMware.CloudFoundation.CertificateManagement.psm1:616]:
```suggestion
            Write-Output "Restarting vCenter Server services for the change to take effect."
            Invoke-RestartService -Service "certificateauthority"
```

**Review by bhumitra — CHANGES_REQUESTED** [review]
Just one more suggestion after what Ryan requested is to add an output message. Rest all good.

File: VMware.CloudFoundation.CertificateManagement.psm1 (hunk 1 above)
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1:655]: addressed in commit 9d2c103e0541e6b077e54e83e2aa0db4e5f07a40

**Comment by garlicNova** [comment]
CHANGELOG.md and .psd1 file updated.

add a helper function.  the suggested method would result in return value contains both the write-output and the return value so i modify it to return an array instead.

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 2]
```diff
@@ -53,6 +53,29 @@ Function Get-Password {
     return $password
 }
 
+function Restart-VcenterService {
```
Comment (bhumitra) [inline VMware.CloudFoundation.CertificateManagement.psm1:56]: Could you please add description for the function and params?
Also, Invoke-RestartService -Service $serviceName -- Do you need any wait or polling ?

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 3]
```diff
@@ -612,8 +635,32 @@ Function Set-EsxiCertificateMode {
...
+                    $failedServices += $serviceStatus.name
+                }
+            }
+
+            if ($failedServices.Count -gt 0) {
+                $failedServicesList = [string]::Join(", ", $failedServices)
+                Write-Error "The following services failed to start: $failedServicesList. Operation Failed." -ErrorAction Stop
+            }
+
+            Invoke-RestartService -Service "certificateauthority"
```
Comment (bhumitra) [inline VMware.CloudFoundation.CertificateManagement.psm1:655]: You will need to change this to call the new function - Restart-VcenterService

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 4]
```diff
@@ -612,8 +635,32 @@ Function Set-EsxiCertificateMode {
...
+            Invoke-RestartService -Service "certificatemanagement"
+            $vmcaStatus = Invoke-GetService -Service "certificateauthority"
```
Comment (bhumitra) [inline VMware.CloudFoundation.CertificateManagement.psm1:657]: This wont be needed as it is part of the function now.

**Review by bhumitra — CHANGES_REQUESTED** [review]
In summary, could you modify the function - 
Set-EsxiCertificateMode to use the Restart-VcenterService - the new one you created. also add documentation for Restart-VcenterService function.

File: VMware.CloudFoundation.CertificateManagement.psm1 (hunk 2 above)
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1:56]: it's an internal function, so i just add a description to it.  There's no wait or polling required.  the Invoke-RestartService will only return when operation is complete.

Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1:56]: address in 30a32aca1c5451e5a0829cfb001061f9b16d0042

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 5]
```diff
@@ -612,8 +635,32 @@ Function Set-EsxiCertificateMode {
...
+                }
+            }
+
+            if ($failedServices.Count -gt 0) {
+                $failedServicesList = [string]::Join(", ", $failedServices)
+                Write-Error "The following services failed to start: $failedServicesList. Operation Failed." -ErrorAction Stop
+            }
+
+            Invoke-RestartService -Service "certificateauthority"
+            Invoke-RestartService -Service "certificatemanagement"
```
[the remark this answers is not in the document]
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1]: address in 30a32aca1c5451e5a0829cfb001061f9b16d0042

File: VMware.CloudFoundation.CertificateManagement.psm1 (hunk 3 above)
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1]: address in 30a32aca1c5451e5a0829cfb001061f9b16d0042

File: VMware.CloudFoundation.CertificateManagement.psm1 (hunk 4 above)
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1]: address in 30a32aca1c5451e5a0829cfb001061f9b16d0042

File: VMware.CloudFoundation.CertificateManagement.psm1 [hunk 6]
```diff
@@ -53,6 +53,33 @@ Function Get-Password {
     return $password
 }
 
+function Restart-VcenterService {
+    <#
+    .DESCRIPTION
+        The Restart-VcenterService restart the vcenter service taken from parameter value and returns an object arrary with the service name, the service status (started or stopped) and the result from restart operation.  
+    #>
```
Comment (bhumitra) [inline VMware.CloudFoundation.CertificateManagement.psm1:60]:
```suggestion
     
     .PARAMETER serviceName
     The name of the service on vCenter Server.
    
    #>
```

**Comment by tenthirtyam** [comment]
I still have some open comments.

File: VMware.CloudFoundation.CertificateManagement.psm1 (hunk 6 above)
Comment (garlicNova) [reply VMware.CloudFoundation.CertificateManagement.psm1:67]: addressed in commit eb037523e0d6150f6f688c31fbde63553b588cdf

**Comment by garlicNova** [comment]
> I still have some open comments.

which ones?  I have commented in https://github.com/vmware/powershell-module-for-vmware-cloud-foundation-certificate-management/pull/110#issuecomment-1949604926 to resolve the first 2 comments.  I do not see any other comments?

**Review by bhumitra — DISMISSED** [review]
LGTM!

**Comment by tenthirtyam** [comment]
Kevin has a couple minor changes to make that we discussed yesterday and we'll delay until those are addressed- mainly around the service status of healthy.

**Review by tenthirtyam — APPROVED** [review]
LGTM!

**Pull request merged**